File: app/modules/knowni-prover/android/src/main/java/expo/modules/knowniprover/KnowniProverModule.kt
// KnowniProverModule.kt: exposes prove() to JavaScript on Android. The work is
// in libknowni_prover.so (prover/src/ffi.rs); an AsyncFunction runs off the UI
// thread, which matters for a call that takes seconds.
package expo.modules.knowniprover

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class KnowniProverModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("KnowniProver")

    AsyncFunction("prove") { inputJson: String, zkeyPath: String, zkeySha256: String ->
      if (loaded) nativeProve(inputJson, zkeyPath, zkeySha256)
      else """{"error":"libknowni_prover.so is not in this build","code":"unsupported"}"""
    }
  }

  companion object {
    // A build without the .so (prover/tools/build-android.sh not run) must
    // still start: the module answers "unsupported" instead of crashing.
    private val loaded: Boolean = try {
      System.loadLibrary("knowni_prover")
      true
    } catch (_: UnsatisfiedLinkError) {
      false
    }

    @JvmStatic
    external fun nativeProve(inputJson: String, zkeyPath: String, zkeySha256: String): String
  }
}

File: app/modules/knowni-prover/ios/KnowniProverModule.swift
// KnowniProverModule.swift: exposes prove() to JavaScript on iOS. The work is
// in KnowniProver.xcframework (prover/src/ffi.rs, built by build-ios.sh); an
// AsyncFunction runs off the main queue, which matters for a call of seconds.
// Without the xcframework (only macOS builds it) the module still compiles and
// answers `unsupported`, which the app already shows as such.

import ExpoModulesCore
#if canImport(KnowniProverFFI)
import KnowniProverFFI
#endif

public class KnowniProverModule: Module {
  public func definition() -> ModuleDefinition {
    Name("KnowniProver")

    AsyncFunction("prove") { (inputJson: String, zkeyPath: String, zkeySha256: String) -> String in
      #if canImport(KnowniProverFFI)
      guard let raw = knowni_prove(inputJson, zkeyPath, zkeySha256) else {
        return "{\"error\":\"the prover returned nothing\"}"
      }
      defer { knowni_free_string(raw) }
      return String(cString: raw)
      #else
      return "{\"error\":\"this build has no native prover\",\"code\":\"unsupported\"}"
      #endif
    }
  }
}
